NPE with URL Rewriting Policy if you call the endpoint directly

Error message:
`{"responseCode":500,"message":"java.lang.NullPointerException error occurred with no message. Refer to logs."}`

Error Log:
```
[ERROR] 2022-02-24 08:37:59.735 [vert.x-eventloop-thread-3] DefaultPolicyErrorWriter - java.lang.NullPointerException error occurred with no message. Refer to logs.
java.lang.NullPointerException: null
	at io.apiman.gateway.engine.policies.URLRewritingPolicy.doApply(URLRewritingPolicy.java:55) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.engine.policies.URLRewritingPolicy.doApply(URLRewritingPolicy.java:37) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.engine.policies.AbstractMappedPolicy.apply(AbstractMappedPolicy.java:87) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.engine.policy.RequestChain.lambda$applyPolicy$0(RequestChain.java:67) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.engine.policy.Chain.executeInPolicyContextLoader(Chain.java:299) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.engine.policy.RequestChain.applyPolicy(RequestChain.java:66) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.engine.policy.Chain.doApply(Chain.java:150) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.engine.impl.ApiRequestExecutorImpl.lambda$execute$5(ApiRequestExecutorImpl.java:301) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.engine.impl.ApiRequestExecutorImpl.lambda$loadPolicies$7(ApiRequestExecutorImpl.java:682) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.engine.policy.PolicyFactoryImpl.loadPolicy(PolicyFactoryImpl.java:95) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.engine.impl.ApiRequestExecutorImpl.loadPolicies(ApiRequestExecutorImpl.java:650) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.engine.impl.ApiRequestExecutorImpl.execute(ApiRequestExecutorImpl.java:371) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.platforms.vertx3.http.HttpPolicyAdapter._execute(HttpPolicyAdapter.java:109) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.platforms.vertx3.http.HttpPolicyAdapter.execute(HttpPolicyAdapter.java:58) ~[apiman-gateway.jar:?]
	at io.apiman.gateway.platforms.vertx3.verticles.HttpGatewayVerticle.requestHandler(HttpGatewayVerticle.java:55) ~[apiman-gateway.jar:?]
	at io.vertx.core.http.impl.WebSocketRequestHandler.handle(WebSocketRequestHandler.java:50) ~[apiman-gateway.jar:?]
	at io.vertx.core.http.impl.WebSocketRequestHandler.handle(WebSocketRequestHandler.java:32) ~[apiman-gateway.jar:?]
	at io.vertx.core.http.impl.Http1xServerConnection.handleMessage(Http1xServerConnection.java:136) ~[apiman-gateway.jar:?]
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366) ~[apiman-gateway.jar:?]
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43) ~[apiman-gateway.jar:?]
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:229) ~[apiman-gateway.jar:?]
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:164) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[apiman-gateway.jar:?]
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93) ~[apiman-gateway.jar:?]
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:99) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[apiman-gateway.jar:?]
	at io.vertx.core.http.impl.Http1xUpgradeToH2CHandler.channelRead(Http1xUpgradeToH2CHandler.java:110) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[apiman-gateway.jar:?]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324) ~[apiman-gateway.jar:?]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[apiman-gateway.jar:?]
	at io.vertx.core.http.impl.Http1xOrH2CHandler.end(Http1xOrH2CHandler.java:61) ~[apiman-gateway.jar:?]
	at io.vertx.core.http.impl.Http1xOrH2CHandler.channelRead(Http1xOrH2CHandler.java:38) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[apiman-gateway.jar:?]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[apiman-gateway.jar:?]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[apiman-gateway.jar:?]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[apiman-gateway.jar:?]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[apiman-gateway.jar:?]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722) ~[apiman-gateway.jar:?]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658) ~[apiman-gateway.jar:?]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584) ~[apiman-gateway.jar:?]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496) ~[apiman-gateway.jar:?]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986) ~[apiman-gateway.jar:?]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[apiman-gateway.jar:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[apiman-gateway.jar:?]

```

Steps to reproduce:

1. Add a URL Rewriting Policy to your API
2. Directly call the URL that is shown you in Apiman in the "Endpoint" tab

--- a/gateway/engine/beans/src/main/java/io/apiman/gateway/engine/beans/ApiRequest.java
+++ b/gateway/engine/beans/src/main/java/io/apiman/gateway/engine/beans/ApiRequest.java
@@ -20,6 +20,8 @@ import io.apiman.gateway.engine.beans.util.QueryMap;
 
 import java.io.Serializable;
 
+import org.jetbrains.annotations.Nullable;
+
 /**
  * An inbound request for a managed API.
  *
@@ -116,6 +118,7 @@ public class ApiRequest implements IApiObject, Serializable {
     /**
      * @return the destination
      */
+    @Nullable
     public String getDestination() {
         return destination;
     }

--- a/gateway/engine/policies/src/main/java/io/apiman/gateway/engine/policies/URLRewritingPolicy.java
+++ b/gateway/engine/policies/src/main/java/io/apiman/gateway/engine/policies/URLRewritingPolicy.java
@@ -52,7 +52,9 @@ public class URLRewritingPolicy extends AbstractMappedDataPolicy<URLRewritingCon
     protected void doApply(ApiRequest request, IPolicyContext context, URLRewritingConfig config,
             IPolicyChain<ApiRequest> chain) {
         if (config.isProcessRequestUrl()) {
-            request.setDestination(request.getDestination().replaceAll(config.getFromRegex(), config.getToReplacement()));
+            if (request.getDestination() != null) {
+                request.setDestination(request.getDestination().replaceAll(config.getFromRegex(), config.getToReplacement()));
+            }
         }
         if (config.isProcessRequestHeaders()) {
             replaceHeaders(config, request.getHeaders());

--- a/gateway/engine/policies/src/test/java/io/apiman/gateway/engine/policies/URLRewritingPolicyTest.java
+++ b/gateway/engine/policies/src/test/java/io/apiman/gateway/engine/policies/URLRewritingPolicyTest.java
@@ -20,8 +20,11 @@ import io.apiman.gateway.engine.beans.ApiRequest;
 import io.apiman.gateway.engine.beans.ApiResponse;
 import io.apiman.test.policies.*;
 
+import org.assertj.core.api.Assertions;
+import org.assertj.core.error.AssertJMultipleFailuresError;
 import org.junit.Assert;
 import org.junit.Test;
+import org.assertj.core.api.Assertions.*;
 
 import java.util.HashMap;
 import java.util.Map;
@@ -91,6 +94,52 @@ public class URLRewritingPolicyTest extends ApimanPolicyTest {
         final HashMap responseMap = new ObjectMapper().readValue(responseBody, HashMap.class);
         Assert.assertEquals("/api-path?foo=bar", responseMap.get("resource"));
         Assert.assertEquals("/another-path?foo=bar", ((Map) responseMap.get("headers")).get("X-Custom-Location"));
+    }
+
+    /**
+     * Handle null resource
+     */
+    @Test
+    @Configuration("{" +
+            "  \"fromRegex\" : \"(.*)\",\n" +
+            "  \"toReplacement\" : \"/$1?foo=bar\",\n" +
+            "  \"processResponseBody\" : false,\n" +
+            "  \"processResponseHeaders\" : false,\n" +
+            "  \"processRequestHeaders\" : true,\n" +
+            "  \"processRequestUrl\" : true\n" +
+            "}")
+    @BackEndApi(EchoBackEndApi.class)
+    public void must_handle_null_request_path() throws Throwable {
+        final PolicyTestRequest request = PolicyTestRequest.build(PolicyTestRequestType.GET, null);
+
+        final PolicyTestResponse response = send(request);
+        final String responseBody = response.body();
+        Assert.assertNotNull(responseBody);
+        final HashMap responseMap = new ObjectMapper().readValue(responseBody, HashMap.class);
+        Assertions.assertThat(responseMap.get("resource")).isNull();
+    }
+
+    /**
+     * Handle empty string resource
+     */
+    @Test
+    @Configuration("{" +
+            "  \"fromRegex\" : \"(.*)\",\n" +
+            "  \"toReplacement\" : \"/$1?foo=bar\",\n" +
+            "  \"processResponseBody\" : false,\n" +
+            "  \"processResponseHeaders\" : false,\n" +
+            "  \"processRequestHeaders\" : true,\n" +
+            "  \"processRequestUrl\" : true\n" +
+            "}")
+    @BackEndApi(EchoBackEndApi.class)
+    public void must_handle_empty_request_path() throws Throwable {
+        final PolicyTestRequest request = PolicyTestRequest.build(PolicyTestRequestType.GET, "");
+
+        final PolicyTestResponse response = send(request);
+        final String responseBody = response.body();
+        Assert.assertNotNull(responseBody);
+        final HashMap responseMap = new ObjectMapper().readValue(responseBody, HashMap.class);
+        Assertions.assertThat(responseMap.get("resource")).isNull();
     }
 
     public static final class URLRewritingTestBackend implements IPolicyTestBackEndApi {

--- a/gateway/engine/policies/src/test/java/io/apiman/gateway/engine/policies/URLRewritingPolicyTest.java
+++ b/gateway/engine/policies/src/test/java/io/apiman/gateway/engine/policies/URLRewritingPolicyTest.java
@@ -139,7 +139,7 @@ public class URLRewritingPolicyTest extends ApimanPolicyTest {
         final String responseBody = response.body();
         Assert.assertNotNull(responseBody);
         final HashMap responseMap = new ObjectMapper().readValue(responseBody, HashMap.class);
-        Assertions.assertThat(responseMap.get("resource")).isNull();
+        Assertions.assertThat(responseMap.get("resource")).isEqualTo("/?foo=bar");
     }
 
     public static final class URLRewritingTestBackend implements IPolicyTestBackEndApi {

